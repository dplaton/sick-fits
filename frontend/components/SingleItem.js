import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import styled from 'styled-components';
import Head from 'next/head';

import Error from './ErrorMessage';

const SingleItemStyles = styled.div`
    max-width: 1280px;
    margin: 2rem auto;
    box-shadow: ${props => props.theme.bs};
    display: grid;
    grid-auto-columns: 1fr;
    grid-auto-flow: column;
    min-height: 800px;
    img {
        width: 100%;
        height: 100%;
        object-fit: contain;
    }
    .details {
        margin: 3rem;
        font-size: 2rem;
    }
`;

export const SINGLE_ITEM_QUERY = gql`
    query SINGLE_ITEM_QUERY($id: ID!) {
        item(where: { id: $id }) {
            title
            description
            price
            largeImage
            image
        }
    }
`;

class SingleItem extends Component {
    render() {
        return (
            <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
                {({ data, error, loading }) => {
                    if (loading) return <p>Loading...</p>;
                    if (error) return <Error error={error} />;
                    if (!data.item) return <p>No data found for item {this.props.id}</p>;
                    const item = data.item;
                    return (
                        <SingleItemStyles>
                            <Head>
                                <title>Sick Fits | {item.title}</title>
                            </Head>
                            <img src={data.item.largeImage} alt={data.item.title} />
                            <div className="details">
                                <h2>Viewing: {item.title}</h2>
                                <p>{item.description}</p>
                            </div>
                        </SingleItemStyles>
                    );
                }}
            </Query>
        );
    }
}

export default SingleItem;
