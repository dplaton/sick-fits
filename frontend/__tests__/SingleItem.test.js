
import { mount } from 'enzyme';
import toJson  from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import SingleItem, {SINGLE_ITEM_QUERY} from '../components/SingleItem';

import { fakeItem } from '../utils/testUtils'; 

describe('<SingleItem />', () => {
    it('renders the data properly', async () => {
        const mocks = [{
            request: { query: SINGLE_ITEM_QUERY, variables: {id: '123'}},
            result:  { data: { item: fakeItem() } }
        }];

        const wrapper = mount(
            <MockedProvider mocks={mocks}> 
                <SingleItem id='123'/>
            </MockedProvider>
        );
        expect(wrapper.text()).toContain('Loading...');
        await wait(0);
        wrapper.update();
        expect(toJson(wrapper.find('h2'))).toMatchSnapshot();
        expect(toJson(wrapper.find('p'))).toMatchSnapshot();
        expect(toJson(wrapper.find('img'))).toMatchSnapshot();
    });

    it ('renders error when the item is missing', async() => {

        const mocks = [{
            request: { query: SINGLE_ITEM_QUERY, variables: {id: '123'}},
            result: {
                errors:[{
                    message: "Item not found"
                }]
            }
        }];

        const wrapper = mount(<MockedProvider mocks={mocks}>
            <SingleItem id='123'/>
        </MockedProvider>)

        await wait(0);
        wrapper.update();
        const errorMessage = wrapper.find('[data-test="graphql-error"]');
        expect(toJson(errorMessage)).toMatchSnapshot();
    })
})