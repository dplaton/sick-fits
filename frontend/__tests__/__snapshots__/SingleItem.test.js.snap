// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> renders error when the item is missing 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    ðŸ›‘
  </strong>
  Item not found
</p>
`;

exports[`<SingleItem /> renders the data properly 1`] = `
<h2>
  Viewing: 
  dogs are best
</h2>
`;

exports[`<SingleItem /> renders the data properly 2`] = `
<p>
  dogs
</p>
`;

exports[`<SingleItem /> renders the data properly 3`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;
